resolver/resolver.cc

```diff
@@ -773,12 +773,19 @@ class ResolveConstantsWalk {
                 }
             }
         } else {
-            if (!job.klass.data(ctx)->addMixin(ctx, resolvedClass, job.mixinIndex)) {
-                if (auto e = ctx.beginError(job.ancestor->loc(), core::errors::Resolver::IncludesNonModule)) {
-                    e.setHeader("Only modules can be `{}`d, but `{}` is a class", job.isInclude ? "include" : "extend",
-                                resolvedClass.show(ctx));
-                    e.addErrorLine(resolvedClass.data(ctx)->loc(), "`{}` defined as a class here",
-                                   resolvedClass.show(ctx));
+            if (ctx.state.newSymbols().includes(job.klass)) {
+                if (!job.klass.data(ctx)->addMixin(ctx, resolvedClass, job.mixinIndex)) {
+                    if (auto e = ctx.beginError(job.ancestor->loc(), core::errors::Resolver::IncludesNonModule)) {
+                        e.setHeader("Only modules can be `{}`d, but `{}` is a class",
+                                    job.isInclude ? "include" : "extend", resolvedClass.show(ctx));
+                        e.addErrorLine(resolvedClass.data(ctx)->loc(), "`{}` defined as a class here",
+                                       resolvedClass.show(ctx));
+                    }
+                }
+            } else {
+                if (auto e = ctx.beginError(job.ancestor->loc(), core::errors::Resolver::ReopeningOlderStratum)) {
+                    e.setHeader("`{}` may not be reopened, as its package has already been processed",
+                                job.klass.show(ctx));
```

Comment:
Should the logic be even more restrictive? I'm hesitant to build features that are a function of the strata, and not a function of explicit dependencies.

The strata change as an implementation detail. For example, I know that you had plans to maybe group more things into one strata, or to split strata into smaller strata, whichever would achieve better throughput.

It seems bad to have a user-visible error message tied to something as flimsy as an implementation detail of strata sizes.

Reply:
That's fair. I had been thinking of this from the point of view of providing the invariant that I'm after, but that feels harder to justify in an error message.

I think we could change this to be that you can only add a mixin to a symbol that your current package owns, which I think is more in the spirit of what we're moving towards with the packager. I think we might need to make a special carve-out for prelude packages, as the ownership of their symbols is strange, but that fits with their existing purpose anyway.